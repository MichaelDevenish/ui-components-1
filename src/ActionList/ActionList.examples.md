#### Simple Action list with header message

```jsx
<div>
   <ActionList
    headerMessage='This is a header message'
    actions={[
      {
       title: 'This is a first item'
      }, 
      {
       title: 'This is a second item',
       description: 'With a description'
      }
    ]}
  />
</div>
```

#### Action list used in callout with default header message

```jsx
<div>
  <Callout
    type='warning'
  >
    Test callout description
    
    <ActionList
      actions={[
        {
         title: 'This is a first item'
        }, 
        {
         title: 'This is a second item',
         description: 'With a description'
        }
      ]}
    />
  </Callout>
</div>
```
