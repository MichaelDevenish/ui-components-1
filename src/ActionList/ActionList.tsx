import React from 'react'
import { isEmpty, map, cloneDeep } from 'lodash'
import {
  ActionDescriptionWrapper,
  ActionLinksWrapper,
  ActionListBodyWrapper,
  ActionListHeaderWrapper,
  ActionTitleWrapper
} from './style'
import { List } from '../List'
import { TextLink } from '../Link'

interface IActionLink {
  linkUrl: string
  linkText: string
  useReactRouter?: boolean
}

interface IAction {
  title: string
  description?: string
  actionLinks?: IActionLink[]
}

interface IActionList {
  headerMessage?: string
  action?: IAction,
  actions?: IAction[]
}

const descriptionForAction = (action: IAction) => {
  if (!action.description) {
    return null
  }

  return (
    <ActionDescriptionWrapper>
      {action.description}
    </ActionDescriptionWrapper>
  )
}

const actionLinksForAction = (action: IAction) => {
  if (!action.actionLinks) {
    return null
  }

  const actionLinks = map(action.actionLinks, (actionLink, index) =>
    <TextLink
      key={`linktext-${index}`}
      href={actionLink.linkUrl}
      className='action-link'
      useReactRouter={actionLink.useReactRouter || false}
    >
      {actionLink.linkText}
    </TextLink>
  )

  return (
    <ActionLinksWrapper>
      {actionLinks}
    </ActionLinksWrapper>
  )
}

class ActionList extends React.PureComponent<IActionList> {
  public static defaultProps: Partial<IActionList> = {
    headerMessage: 'Actions you can take',
    actions: []
  }

  get actions (): IAction[] {
    const { action } = this.props
    const { actions } = this.props

    let actionsCopy: IAction[] = []

    if (actions) {
      actionsCopy = cloneDeep(actions)
    }

    if (action) {
      actionsCopy.push(action)
    }

    return actionsCopy
  }

  get actionsList (): JSX.Element {
    const actionsList = map(this.actions, (action, index) =>
      <div key={`action-${index}`}>
        <ActionTitleWrapper>
          {action.title}
        </ActionTitleWrapper>

        {descriptionForAction(action)}
        {actionLinksForAction(action)}
      </div>
    )

    return (
      <List type='ordered'>
        {actionsList}
      </List>
    )
  }

  get headerMessage (): JSX.Element | null {
    const {
      headerMessage
    } = this.props

    if (!headerMessage) {
      return null
    }

    return (
      <ActionListHeaderWrapper>
        {headerMessage}
      </ActionListHeaderWrapper>
    )
  }

  public render (): JSX.Element | null {
    if (isEmpty(this.actions)) {
      return null
    }

    return (
      <div className='action-list'>
        {this.headerMessage}

        <ActionListBodyWrapper>
          {this.actionsList}
        </ActionListBodyWrapper>
      </div>
    )
  }
}

export {
  ActionList
}
