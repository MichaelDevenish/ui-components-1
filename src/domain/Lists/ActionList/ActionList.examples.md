#### Simple Action list with header message

```jsx
<div>
   <ActionList
    headerMessage='This is a header message'
    actions={[
      {
       title: 'This is a first item'
      }, 
      {
       title: 'This is a second item',
       description: 'With a description'
      }
    ]}
  />
</div>
```

#### Action list with link

```jsx
<div>
   <ActionList
    headerMessage='This is a header message'
    action={{
      title: 'Test Title',
      description: 'Test Description',
      actionLinks: [
        {
          linkText: 'This is a link',
          linkUrl: '/'
        },
        {
          linkText: 'This is a link with some other props',
          linkUrl: '/',
          anchorComponentProps: {
            someProp: true
          }
        }
      ]
    }}
  />
</div>
```

#### Action list used in callout with default header message

```jsx
import { Callout } from '@Domain/Callouts';

<div>
  <Callout
    type='warning'
  >
    Test callout description
    
    <ActionList
      actions={[
        {
         title: 'This is a first item'
        }, 
        {
         title: 'This is a second item',
         description: 'With a description'
        }
      ]}
    />
  </Callout>
</div>
```


