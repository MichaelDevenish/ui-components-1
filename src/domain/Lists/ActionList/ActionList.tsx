import cloneDeep from 'lodash/cloneDeep'
import isEmpty from 'lodash/isEmpty'
import map from 'lodash/map'
import React from 'react'

import { TextLink } from '../../Links'
import { List } from '../List'
import {
  ActionDescriptionWrapper,
  ActionLinksWrapper,
  ActionListBodyWrapper,
  ActionListHeaderWrapper,
  ActionTitleWrapper
} from './style'

interface IActionLink {
  linkUrl: string
  linkText: string
  anchorComponentProps?: {
    [i: string]: any
  }
}

interface IAction {
  title: string
  description?: string
  actionLinks?: IActionLink[]
}

interface IActionList {
  headerMessage?: string
  action?: IAction,
  actions?: IAction[]
}

class ActionList extends React.PureComponent<IActionList> {
  get actions (): IAction[] {
    const {
      action,
      actions
    } = this.props

    let actionsCopy: IAction[] = []

    if (actions) {
      actionsCopy = cloneDeep(actions)
    }

    if (action) {
      actionsCopy.push(action)
    }

    return actionsCopy
  }

  get actionsList (): JSX.Element {
    const actionsList = map(this.actions, (action, index) => (
      <div key={`action-${index}`}>
        <ActionTitleWrapper>
          {action.title}
        </ActionTitleWrapper>

        {this.descriptionForAction(action.description)}
        {this.actionLinksForAction(action.actionLinks)}
      </div>
    ))

    return (
      <List type='ordered'>
        {actionsList}
      </List>
    )
  }

  get headerMessage (): JSX.Element | null {
    const {
      headerMessage
    } = this.props

    if (!headerMessage) {
      return null
    }

    return (
      <ActionListHeaderWrapper>
        {headerMessage}
      </ActionListHeaderWrapper>
    )
  }
  public static defaultProps: Partial<IActionList> = {
    headerMessage: 'Actions you can take',
    actions: []
  }

  public render (): JSX.Element | null {
    if (isEmpty(this.actions)) {
      return null
    }

    return (
      <div className='action-list'>
        {this.headerMessage}

        <ActionListBodyWrapper>
          {this.actionsList}
        </ActionListBodyWrapper>
      </div>
    )
  }

  private descriptionForAction (description?: string) {
    if (!description) {
      return null
    }

    return (
      <ActionDescriptionWrapper>
        {description}
      </ActionDescriptionWrapper>
    )
  }

  private actionLinksForAction (actionLinks?: IActionLink[]) {
    if (!actionLinks) {
      return null
    }

    const textLinkComponents = map(actionLinks, (actionLink, index) => (
      <TextLink
        key={`action-link-${index}`}
        href={actionLink.linkUrl}
        className='action-link'
        anchorComponentProps={actionLink.anchorComponentProps}
      >
        {actionLink.linkText}
      </TextLink>
    ))

    return (
      <ActionLinksWrapper>
        {textLinkComponents}
      </ActionLinksWrapper>
    )
  }
}

export {
  ActionList,
  IActionList,
  IActionLink,
  IAction
}
