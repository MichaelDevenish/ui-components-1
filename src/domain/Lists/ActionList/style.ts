import styled from 'styled-components'

const ActionDescriptionWrapper = styled.div`
  margin-top: 6px;
`

const ActionLinksWrapper = styled.div`
  margin-top: 14px;
  
  .action-link {
    display: block;
    margin-top: 0;
    margin-bottom: 8px;
  }
`

const ActionTitleWrapper = styled.div`
  font-weight: 800;
`

const ActionListHeaderWrapper = styled.div`
  margin-bottom: 10px;
`

const ActionListBodyWrapper = styled.div`
  ol {
    counter-reset: ol-action-list;
    margin-left:0;
    padding-left:0;
    list-style: none;

    > li {
      display: flex;
      margin-bottom: 16px;

      &:last-child {
        margin-bottom: 0;
      }

      &:before {
        content: counter(ol-action-list) '.';
        counter-increment: ol-action-list;
        font-weight: 800;
        margin-right: 26px;
      }
    }
  }
`

export {
  ActionTitleWrapper,
  ActionLinksWrapper,
  ActionDescriptionWrapper,
  ActionListHeaderWrapper,
  ActionListBodyWrapper
}
