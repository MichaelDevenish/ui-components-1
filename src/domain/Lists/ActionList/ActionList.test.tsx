import React from 'react'
import { shallow } from 'enzyme'
import { ActionList } from './ActionList'

describe('<ActionList />', () => {
  it('should render a simple action list with multiple actions', () => {
    const wrapper = shallow(
      <ActionList
        actions={[
          {
            title: 'Test Title 1'
          },
          {
            title: 'Test Title 2',
            description: 'Test description'
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a simple action list with a single action', () => {
    const wrapper = shallow(
      <ActionList
        action={{
          title: 'Contact our customer support',
          description: 'Test description'
        }}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a simple action list with a header', () => {
    const wrapper = shallow(
      <ActionList
        headerMessage='Test header message'
        action={{
          title: 'Contact our customer support',
          description: 'Test description'
        }}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a simple action list with a link', () => {
    const wrapper = shallow(
      <ActionList
        headerMessage='Test header message'
        action={{
          title: 'Test Title',
          description: 'Test description',
          actionLinks: [
            {
              linkText: 'Test Link Text',
              linkUrl: '/test-url'
            }
          ]
        }}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
