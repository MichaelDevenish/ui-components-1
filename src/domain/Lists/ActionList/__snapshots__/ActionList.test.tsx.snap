// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionList /> should render a simple action list with a header 1`] = `
<div
  className="action-list"
>
  <styled.div>
    Test header message
  </styled.div>
  <styled.div>
    <List
      type="ordered"
    >
      <div
        key="action-0"
      >
        <styled.div>
          Contact our customer support
        </styled.div>
        <styled.div>
          Test description
        </styled.div>
      </div>
    </List>
  </styled.div>
</div>
`;

exports[`<ActionList /> should render a simple action list with a link 1`] = `
<div
  className="action-list"
>
  <styled.div>
    Test header message
  </styled.div>
  <styled.div>
    <List
      type="ordered"
    >
      <div
        key="action-0"
      >
        <styled.div>
          Test Title
        </styled.div>
        <styled.div>
          Test description
        </styled.div>
        <styled.div>
          <WithAnchor
            className="action-link"
            href="/test-url"
            key="action-link-0"
            useReactRouter={false}
          >
            Test Link Text
          </WithAnchor>
        </styled.div>
      </div>
    </List>
  </styled.div>
</div>
`;

exports[`<ActionList /> should render a simple action list with a single action 1`] = `
<div
  className="action-list"
>
  <styled.div>
    Actions you can take
  </styled.div>
  <styled.div>
    <List
      type="ordered"
    >
      <div
        key="action-0"
      >
        <styled.div>
          Contact our customer support
        </styled.div>
        <styled.div>
          Test description
        </styled.div>
      </div>
    </List>
  </styled.div>
</div>
`;

exports[`<ActionList /> should render a simple action list with multiple actions 1`] = `
<div
  className="action-list"
>
  <styled.div>
    Actions you can take
  </styled.div>
  <styled.div>
    <List
      type="ordered"
    >
      <div
        key="action-0"
      >
        <styled.div>
          Test Title 1
        </styled.div>
      </div>
      <div
        key="action-1"
      >
        <styled.div>
          Test Title 2
        </styled.div>
        <styled.div>
          Test description
        </styled.div>
      </div>
    </List>
  </styled.div>
</div>
`;
